lib/wasi/src/syscalls/windows.rs

```diff
@@ -5,42 +5,40 @@ use wasmer::WasmCell;
 pub fn platform_clock_res_get(
     clock_id: __wasi_clockid_t,
     resolution: WasmCell<__wasi_timestamp_t>,
-) -> __wasi_errno_t {
+) -> Result<__wasi_timestamp_t, __wasi_errno_t> {
     let resolution_val = match clock_id {
         // resolution of monotonic clock at 10ms, from:
         // https://docs.microsoft.com/en-us/windows/desktop/api/sysinfoapi/nf-sysinfoapi-gettickcount64
         __WASI_CLOCK_MONOTONIC => 10_000_000,
         // TODO: verify or compute this
         __WASI_CLOCK_REALTIME => 1,
         __WASI_CLOCK_PROCESS_CPUTIME_ID => {
-            return __WASI_EINVAL;
+            return Err(__WASI_EINVAL);
         }
         __WASI_CLOCK_THREAD_CPUTIME_ID => {
-            return __WASI_EINVAL;
+            return Err(__WASI_EINVAL);
         }
-        _ => return __WASI_EINVAL,
+        _ => return Err(__WASI_EINVAL),
     };
-    resolution.set(resolution_val);
-    __WASI_ESUCCESS
+    Ok(resolution_val)
 }
 
 pub fn platform_clock_time_get(
     clock_id: __wasi_clockid_t,
     precision: __wasi_timestamp_t,
-    time: WasmCell<__wasi_timestamp_t>,
-) -> __wasi_errno_t {
+) -> Result<__wasi_timestamp_t, __wasi_errno_t> {
```

Comment:
This is changing the spec of the WASI `platform_clock_time_get` signature. Did it got updated and we missed it?

Reply:
My bad - I should not have changed this file - will put it back

Reply:
So looked into it and what I'd done is to hide more of the details of the WASI syscall from the proxies:

Basically the memory copies to WASM memory are done sooner than the implementations themselves...
https://github.com/john-sharratt/wasmer/blob/master/lib/wasi/src/syscalls/mod.rs#L82

Really I guess the choice is do you allow consumers to proxy or not?

If the answer is no... then I could look into using the ChainedNamedResolver as you suggested.
If the answer is yes... then I could look into reducing the abstraction leakage as much as possible and moving the memory copies into a nice place so that they don't bloat the code.

Not sure there is a right answer and its your hard work so just let me know how I can help.

Cheers

Reply:
> If the answer is yes... then I could look into reducing the abstraction leakage as much as possible and moving the memory copies into a nice place so that they don't bloat the code.

That sounds like a great tradeoff, if we can avoid the proxy abstraction and use ChainedResolver instead I think that will simplify things

> I've done and polling functionality thats not yet implemented on your mainline but needed for more advanced programs like wapm itself

I'd love to have that upstream as well (even the experimental wasi-net)... I really love the concept of having a self-hosted wapm published in wapm 😊